### PR TITLE
added selector engine

### PR DESCRIPTION
**Pull request by mschop**
@fabianruette ready for review

### CODE REVIEW

File: src/Selector.php [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    private static function getPaths($node, $rules)
+    {
+        $result;
+        if($rules['type'] === 'ruleSet') {
+            $result = static::walkNode($rules['rule'], $node, [new PathStep(null, $node)]);
+        } else {
+            $allResults = array_map(function ($ruleSet) use ($node) {
+                return static::walkNode($ruleSet['rule'], $node, [new PathStep(null, $node)]);
+            }, $rules['selectors']);
+            $result = call_user_func_array('array_merge', $allResults);
```
Comment (fabianruette) [inline src/Selector.php]: use array_reduce to clear up intent and avoid call_user_func_array

File: src/Selector.php [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            $idenfication = static::getIdentification($path);
+            if(!in_array($idenfication, $known)) {
+                $known[] = $idenfication;
+                $result[] = $path;
+            }
+        }
+        return $result;
+    }
+
+    private static function getIdentification($path)
```
Comment (fabianruette) [inline src/Selector.php:53]: this could use array_reduce as well

File: src/Selector.php (hunk 1 above)
Comment (mschop) [inline src/Selector.php]: @fabianruette ready for review

File: src/Selector.php (hunk 2 above)
Comment (mschop) [inline src/Selector.php:53]: @fabianruette ready for review

**Pull request merged**